golem-cli/src/api_definition.rs

```diff
@@ -0,0 +1,290 @@
+    #[command()]
+    Delete {
+        /// Api definition id
+        #[arg(short, long)]
+        id: ApiDefinitionId,
+
+        /// Version of the api definition
+        #[arg(short = 'V', long)]
+        version: ApiDefinitionVersion,
+    },
+}
+
+#[async_trait]
+pub trait ApiDefinitionHandler {
+    async fn handle(&self, subcommand: ApiDefinitionSubcommand) -> Result<GolemResult, GolemError>;
+}
+
+pub struct ApiDefinitionHandlerLive<C: ApiDefinitionClient + Send + Sync> {
+    pub client: C,
+}
+
+#[async_trait]
+impl<C: ApiDefinitionClient + Send + Sync> ApiDefinitionHandler for ApiDefinitionHandlerLive<C> {
+    async fn handle(&self, subcommand: ApiDefinitionSubcommand) -> Result<GolemResult, GolemError> {
+        match subcommand {
+            ApiDefinitionSubcommand::Get { id, version } => {
+                let definition = self.client.get(id, version).await?;
+                Ok(GolemResult::Ok(Box::new(definition)))
+            }
+            ApiDefinitionSubcommand::Put { definition } => {
+                let definition = self.client.put(definition).await?;
+                Ok(GolemResult::Ok(Box::new(definition)))
+            }
+            ApiDefinitionSubcommand::List { .. } => {
+                let definitions = self.client.all_get().await?;
+                Ok(GolemResult::Ok(Box::new(definitions)))
+            }
+            ApiDefinitionSubcommand::Delete { id, version } => {
+                let result = self.client.delete(id, version).await?;
+                Ok(GolemResult::Str(result))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::sync::{Arc, Mutex};
+
+    use futures_util::FutureExt;
+    use golem_client::{api::ApiDefinitionError, model::HttpApiDefinition};
+
+    use crate::{
+        api_definition::ApiDefinitionSubcommand,
+        clients::api_definition::ApiDefinitionClientLive,
+        model::{ApiDefinitionId, ApiDefinitionVersion, GolemError},
+    };
+    use golem_client::Error;
+
+    use super::{ApiDefinitionHandler, ApiDefinitionHandlerLive};
+
+    pub struct ApiDefinitionClientTest {
+        calls: Arc<Mutex<String>>,
+    }
+
+    impl golem_client::api::ApiDefinitionClient for ApiDefinitionClientTest {
+        #[must_use]
+        #[allow(clippy::type_complexity, clippy::type_repetition_in_bounds)]
+        fn oas_put<'life0, 'async_trait>(
+            &'life0 self,
+        ) -> ::core::pin::Pin<
+            Box<
+                dyn ::core::future::Future<
+                        Output = Result<HttpApiDefinition, Error<ApiDefinitionError>>,
+                    > + ::core::marker::Send
+                    + 'async_trait,
+            >,
+        >
+        where
+            'life0: 'async_trait,
+            Self: 'async_trait,
+        {
+            todo!()
+        }
+
+        #[must_use]
+        #[allow(clippy::type_complexity, clippy::type_repetition_in_bounds)]
+        fn get<'life0, 'life1, 'life2, 'async_trait>(
```

Comment:
You don't need these horrible function signatures - this is the desugared version of how async functions in a trait look like. Please use the `async_trait` annotation (as you did in other places) and just write async functions instead - this is very hard to read

Reply:
indeed, thanks! I was wondering what is missing to have the simplified signature 🤦